File: Power Query/300-399/397/PQ_397.R
library(tidyverse)
library(readxl)

path <- "300-399/397/PQ_Challenge_397.xlsx"
input <- read_excel(path, range = "A1:C30") %>% na.omit()
test <- read_excel(path, range = "E1:F3")

final_map <- unique(input$Month) %>%
  reduce(
    function(o, m) {
      x <- filter(input, Month == m)
      o[o == o[pull(slice_min(x, Revenue, n = 1), Territory)]] <- o[pull(
        slice_max(x, Revenue, n = 1),
        Territory
      )]
      o
    },
    .init = set_names(unique(input$Territory), unique(input$Territory))
  )

result <- tibble(Territory = names(final_map), Owner = unname(final_map)) %>%
  summarise(Territories = str_c(Territory, collapse = ", "), .by = Owner)

all.equal(result, test)
## [1] TRUE

File: Power Query/300-399/397/PQ_397.py
import pandas as pd

path = "300-399/397/PQ_Challenge_397.xlsx"
input = pd.read_excel(path, usecols="A:C", nrows=30, skiprows=0)
test = pd.read_excel(path, usecols="E:F", nrows=2, skiprows=0)

owner = {t: t for t in input["Territory"].unique()}

for _, g in input.groupby("Month", sort=False):
    win = g.loc[g["Revenue"].idxmax(), "Territory"]
    lose = g.loc[g["Revenue"].idxmin(), "Territory"]

    old, new = owner[lose], owner[win]
    owner = {t: new if o == old else o for t, o in owner.items()}

output = (
    pd.Series(owner, name="Owner")
    .rename_axis("Territory")
    .reset_index()
    .groupby("Owner", sort=False)["Territory"]
    .agg(", ".join)
    .reset_index(name="Territories")
)
print(output.equals(test))
# True